Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
     }
 }
 
+AccessibilityObject* AccessibilityNodeObject::parentTable() const
+{
+    RefPtr element = this->element();
+    if (!element && !AXTableHelpers::isTableRowElement(*element))
+        return nullptr;
+
+    // The parent table might not be the direct ancestor of the row unfortunately. ARIA states that role="grid" should
+    // only have "row" elements, but if not, we still should handle it gracefully by finding the right table.
+    for (RefPtr parent = parentObject(); parent; parent = parent->parentObject()) {
+        if (parent->isTable()) {
+            bool isNonGridRowOrValidAriaTable = !isARIAGridRow() || parent->isAriaTable() || elementName() == ElementName::HTML_tr;
+            if (parent->isExposableTable() && isNonGridRowOrValidAriaTable)
+                return parent.get();
+
+            // If this is a non-anonymous table object, but not an accessibility table, we should stop because we don't want to
+            // choose another ancestor table as this row's table.
+            // Don't exit for ARIA grids, since they could have <table>'s between rows and the owning grid (see aria-grid-with-strange-hierarchy.html).
+            if (!isARIAGridRow() && parent->node())
+                break;
+        }
+    }
+    return nullptr;
+}
+
+void AccessibilityNodeObject::setRowIndex(unsigned rowIndex)
+{
+    if (role() != AccessibilityRole::Row)
+        return;
+
+    CheckedRef rareData = ensureRareData();
+    if (rareData->rowIndex() == rowIndex)
+        return;
+    rareData->setRowIndex(rowIndex);
+
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+    if (CheckedPtr cache = axObjectCache())
+        cache->rowIndexChanged(*this);
+#endif
+}
+
+std::optional<unsigned> AccessibilityNodeObject::axColumnIndex() const
+{
+    if (role() != AccessibilityRole::Row)
+        return std::nullopt;
+
+    int value = integralAttribute(aria_colindexAttr);
+    return value >= 1 ? std::optional(value) : std::nullopt;
+}
+
+std::optional<unsigned> AccessibilityNodeObject::axRowIndex() const
+{
+    if (role() != AccessibilityRole::Row)
+        return std::nullopt;
+
+    int value = integralAttribute(aria_rowindexAttr);
+    return value >= 1 ? std::optional(value) : std::nullopt;
+}
+
+String AccessibilityNodeObject::axRowIndexText() const
+{
+    return getAttribute(aria_rowindextextAttr);
+}
+
+AccessibilityObject::AccessibilityChildrenVector AccessibilityNodeObject::disclosedRows()
+{
+    if (!isARIATreeGridRow())
+        return AccessibilityObject::disclosedRows();
+
+    AccessibilityChildrenVector disclosedRows;
+
+    // The contiguous disclosed rows will be the rows in the table that
+    // have an aria-level of plus 1 from this row.
+    Ref parent = *parentObjectUnignored();
+    if (!parent->isExposableTable())
+        return disclosedRows;
```

Comment:
I know you didn't add this code, but the unconditional deref seems unnecessary. Maybe we can fix it while we're here?

```cpp
RefPtr parent = parentObjectUnignored();
if (!parent || !parent->isExposableTable())
    return { };

AccessibilityChildrenVector disclosedRows;
```